File: src/main/java/at/michael1011/backpacks/listeners/Join.java
package at.michael1011.backpacks.listeners;

import at.michael1011.backpacks.Main;
import at.michael1011.backpacks.SQL;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;

import java.sql.ResultSet;
import java.sql.SQLException;

public class Join implements Listener {

    public Join(Main main) {
        main.getServer().getPluginManager().registerEvents(this, main);
    }

    // fixme: send update message to all players with the permission 'backpacks.update'

    @EventHandler(priority = EventPriority.HIGH)
    public void joinEvent(PlayerJoinEvent e) {
        final Player p = e.getPlayer();
        final String playerName = p.getName();

        if(!Main.availablePlayers.contains(playerName)) {
            Main.availablePlayers.add(playerName);
        }

        final String uuid = p.getUniqueId().toString().replaceAll("-", "");

        SQL.getResult("SELECT * FROM bp_users WHERE uuid='"+uuid+"'", new SQL.Callback<ResultSet>() {
            @Override
            public void onSuccess(ResultSet rs) {
                try {
                    if (rs != null) {
                        if(rs.first()) {
                            if(!rs.getString("name").equals(playerName)) {
                                SQL.query("UPDATE bp_users SET name='"+playerName+"' WHERE uuid='"+uuid+"'", new SQL.Callback<Boolean>() {
                                    @Override
                                    public void onSuccess(Boolean rs) {}

                                    @Override
                                    public void onFailure(Throwable e) {}

                                });
                            }

                        } else {
                            SQL.query("INSERT INTO bp_users (name, uuid) VALUES ('"+playerName+"', '"+uuid+"')", new SQL.Callback<Boolean>() {
                                @Override
                                public void onSuccess(Boolean rs) {}

                                @Override
                                public void onFailure(Throwable e) {}

                            });
                        }

                        rs.close();
                    }

                } catch (SQLException ex) {
                    ex.printStackTrace();
                }

            }

            @Override
            public void onFailure(Throwable e) {}

        });

    }

}

File: src/main/kotlin/at/michael1011/backpacks/listeners/Join.kt
package at.michael1011.backpacks.listeners

import at.michael1011.backpacks.database.Database
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent

class Join(private val database: Database) : Listener {
    // In case of a reload the database also has to be updated
    init {
        Bukkit.getServer().onlinePlayers.forEach {
            updatePlayer(it)
        }
    }

    @EventHandler(priority = EventPriority.NORMAL)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        updatePlayer(event.player)
    }

    private fun updatePlayer(player: Player) {
        // Insert a new player or just update the name if the UUID is already in the database
        // The exact queries vary from SQL implementation therefore we need two different ones for SQLite and MySQL
        val statement = if (database.isMysql) {
            database.connection.prepareStatement(
                "INSERT INTO players VALUES (?, ?) ON DUPLICATE KEY UPDATE name = ?"
            )
        } else {
            database.connection.prepareStatement(
                "INSERT INTO players VALUES (?, ?) ON CONFLICT(uuid) DO UPDATE SET name = ?"
            )
        }

        statement.setString(1, player.uniqueId.toString())
        statement.setString(2, player.name)
        statement.setString(3, player.name)

        database.executeStatement(statement) {}
    }
}
